### PR TITLE
Firts pull req second to master

### PR DESCRIPTION
**Pull request by Bronzza**

### CODE REVIEW

File: .idea/uiDesigner.xml [hunk 1]
```diff
@@ -0,0 +1,124 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Anderwerty) [inline .idea/uiDesigner.xml]: This file should be in the repository.

File: src/main/java/MyArrayList.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
+public class MyArrayList<T> implements MyList<T> {
```
Comment (Anderwerty) [inline src/main/java/MyArrayList.java]: Please use some root package. Don't put classes just inti java package.

File: src/main/java/MyArrayList.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    private Object[] array;
+    private int counter = 0;
+    private final int DEFAULT_CAPACITY = 10;
```
Comment (Anderwerty) [inline src/main/java/MyArrayList.java]: It is not the right declaration.  You forget modifier static.

File: src/main/java/MyArrayList.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+    public MyArrayList() {
+        array = new Object[DEFAULT_CAPACITY];
+    }
+
```
Comment (Anderwerty) [inline src/main/java/MyArrayList.java]: Remove redundant blank line.

File: src/main/java/MyArrayList.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+    @Override
+    public T get(int i) {
+        for (int j = 0; j < array.length; j++) {
```
Comment (Anderwerty) [inline src/main/java/MyArrayList.java]: The complexity of your algorithm is O(N) but should be O(1).

File: src/main/java/MyArrayList.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                sb.append(array[i] + ",");
+            } else {
+                sb.append(array[i]);
+            }
+        }
+        sb.append("]");
+        return sb.toString();
+    }
+
+    private boolean checkIsExpand() {
```
Comment (Anderwerty) [inline src/main/java/MyArrayList.java]: Implementation is not right.

File: src/main/java/MyArrayList.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        for (int j = 0; j < array.length; j++) {
+            if (j == i) {
+                return (T) array[j];
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public void add(T t) {
```
Comment (Anderwerty) [inline src/main/java/MyArrayList.java]: What if counter>DEFAULT_CAPACITY? Do you have a test for this scenario?

File: src/main/java/MyArrayList.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        array[counter] = t;
+        counter++;
+        if (checkIsExpand()) {
+            expandArray();
+        }
+    }
+
+
+    public void add(int index, T t) {
+        if (index < 0 || index > counter) {
```
Comment (Anderwerty) [inline src/main/java/MyArrayList.java]: Is it ok if index=counter.?
Please move validation into a separate method and reuse it.

File: src/main/java/MyArrayList.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+            System.arraycopy(array, 0, temp, 0, index);
+            temp[index] = t;
+            System.arraycopy(array, index, temp, index + 1, counter - index);
+            array = temp;
+            counter++;
+        }
+        if (checkIsExpand()) {
+            expandArray();
+        }
+
```
Comment (Anderwerty) [inline src/main/java/MyArrayList.java]: Remove redundant blank line.

File: src/main/java/MyArrayList.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        }
+    }
+
+    @Override
+    public int size() {
+        return counter;
+    }
+
+    @Override
+    public String toString() {
```
Comment (Anderwerty) [inline src/main/java/MyArrayList.java]: You could use Arrays.toString()

File: src/main/java/MyLinkedList.java [hunk 11]
```diff
@@ -0,0 +1,159 @@
+public class MyLinkedList<T> implements MyList<T> {
```
Comment (Anderwerty) [inline src/main/java/MyLinkedList.java]: The same

File: src/main/java/MyLinkedList.java [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+            this.prev = prev;
+        }
+    }
+
+    @Override
+    public T get(int i) {
+        if (checkIndex(i)) {
+            return findElement(i).value;
+        } else {
+            throw new IndexOutOfBoundsException();
```
Comment (Anderwerty) [inline src/main/java/MyLinkedList.java]: Please add a test for this scenario.

File: src/main/java/MyLinkedList.java [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+            } else {
+                Node indexNode = findElement(index);
+                Node temp = indexNode.prev;
+                Node<T> createdNode = new Node(temp, element, indexNode);
+                indexNode.prev = createdNode;
+                temp.next = createdNode;
+                size++;
+            }
+        } else {
+            throw new IndexOutOfBoundsException();
```
Comment (Anderwerty) [inline src/main/java/MyLinkedList.java]: The same

File: src/main/java/MyLinkedList.java [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+            last = createdNode;
+        } else {
+            temp.prev = createdNode;
+        }
+        size++;
+    }
+
+    private T unlink(Node<T> node) {
+        if (node == null) {
+            throw new NullPointerException();
```
Comment (Anderwerty) [inline src/main/java/MyLinkedList.java]: NullPointerException could be thrown by the code below. So it is better to create your own custom exception and throw it.

File: src/main/java/MyLinkedList.java [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+            sb.append(tmp.value);
+            while (tmp.next != null) {
+                sb.append(", ");
+                tmp = tmp.next;
+                sb.append(tmp.value);
+            }
+            return sb.toString();
+        }
+    }
+
```
Comment (Anderwerty) [inline src/main/java/MyLinkedList.java]: Remove redundant blank lines.

File: src/main/java/MyList.java [hunk 16]
```diff
@@ -0,0 +1,9 @@
+public interface MyList<T> {
```
Comment (Anderwerty) [inline src/main/java/MyList.java]: the same

File: src/main/java/MyStack.java [hunk 17]
```diff
@@ -0,0 +1,21 @@
+public class MyStack<T> extends MyLinkedList implements MyStackInterface <T> {
```
Comment (Anderwerty) [inline src/main/java/MyStack.java]: The same. And why don't use generic for MyLinkedList?

File: src/main/java/MyStack.java [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+        T value = (T) get(0);
+        remove(0);
+        return value;
+    }
+
+    @Override
+    public T peek() {
+        return (T) get(0);
+    }
+
```
Comment (Anderwerty) [inline src/main/java/MyStack.java]: Remove the blank line.

File: src/main/java/MyStackInterface.java [hunk 19]
```diff
@@ -0,0 +1,8 @@
+public interface MyStackInterface <T> {
```
Comment (Anderwerty) [inline src/main/java/MyStackInterface.java]: The same

File: src/main/java/MyStackInterface.java [hunk 20]
```diff
@@ -0,0 +1,8 @@
...
+    public T push(T element);
+
+    public T pop();
+
+    public T peek();
+
```
Comment (Anderwerty) [inline src/main/java/MyStackInterface.java]: Remove redundant blank line.

File: src/test/java/MyLinkedListTest.java [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Stream;
+
+import static org.junit.Assert.*;
```
Comment (Anderwerty) [inline src/test/java/MyLinkedListTest.java]: Don't use wild cards in imports.

File: src/test/java/MyLinkedListTest.java [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+        assertEquals(Exception.class, myLinkedList.get(myLinkedList.size()));
+    }
+
+    @Test
+    public void add() {
+        myLinkedList.add(value);
+        assertEquals(value, myLinkedList.get(myLinkedList.size() - 1));
+        myLinkedList.add(null);
+        assertNull(myLinkedList.get(myLinkedList.size()-1));
+
```
Comment (Anderwerty) [inline src/test/java/MyLinkedListTest.java]: Remove redundant blank lines.

File: src/test/java/MyLinkedListTest.java [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+    @Test
+    public void add1() throws IndexOutOfBoundsException {
+        int index = randomInt(myLinkedList.size());
+        myLinkedList.add(index, value);
+        assertEquals(value, myLinkedList.get(index));
+        trown.expect(IndexOutOfBoundsException.class);
+        myLinkedList.add(-1, value);
+        trown.expect(IndexOutOfBoundsException.class);
+        myLinkedList.add(myLinkedList.size()+1, value);
+
```
Comment (Anderwerty) [inline src/test/java/MyLinkedListTest.java]: The same

File: src/test/java/MyStackTestAbstract.java [hunk 24]
```diff
@@ -0,0 +1,8 @@
+public abstract class MyStackTestAbstract <T> {
+    protected MyStack<T> myStack;
+    protected T value;
+
+    public abstract T createSampleValue();
+    public abstract MyStack<T> createSampleStack();
+
+}
```
Comment (Anderwerty) [inline src/test/java/MyStackTestAbstract.java]: Add a blank line at the end of the file.

File: src/test/java/MyStackTest.java [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+    @Override
+    public String createSampleValue() {
+        return "String";
+    }
+
+    @Override
+    public MyStack<String> createSampleStack() {
+        return new MyStack<>();
+    }
+}
```
Comment (Anderwerty) [inline src/test/java/MyStackTest.java]: Add a blank line at the end of the file.

File: src/test/java/MyStackTest.java [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+
+    @Test
+    public void push() {
+        assertEquals(value,myStack.peek());
+    }
+
+    @Test (expected = IndexOutOfBoundsException.class)
+    public void pop() {
+        myStack.pop();
+        assertEquals(Exception.class, myStack.pop());
```
Comment (Anderwerty) [inline src/test/java/MyStackTest.java]: This assert does not work. `assertEquals(Exception.class, myStack.pop());->myStack.pop();`

File: src/test/java/MyStackTest.java [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+
+import static org.junit.Assert.assertEquals;
+
+public class MyStackTest extends MyStackTestAbstract <String> {
+
+    @Before
+    public void setUp(){
+        myStack = createSampleStack();
+        value = createSampleValue();
+        myStack.push(value);
```
Comment (Anderwerty) [inline src/test/java/MyStackTest.java]: Why you push only one element?

File: src/test/java/MyStackTest.java [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @Test
+    public void push() {
+        assertEquals(value,myStack.peek());
```
Comment (Anderwerty) [inline src/test/java/MyStackTest.java]: Please format your code.

File: src/test/java/MyStackTest.java [hunk 29]
```diff
@@ -0,0 +1,42 @@
...
+
+    @Test (expected = IndexOutOfBoundsException.class)
+    public void pop() {
+        myStack.pop();
+        assertEquals(Exception.class, myStack.pop());
+    }
+
+    @Test
+    public void peek() {
+        myStack.peek();
```
Comment (Anderwerty) [inline src/test/java/MyStackTest.java]: Why you don't use assert for this line?

File: src/test/java/MyStackTest.java [hunk 30]
```diff
@@ -0,0 +1,42 @@
...
+        assertEquals(value, myStack.peek());
+    }
+
+
+    @Override
+    public String createSampleValue() {
+        return "String";
```
Comment (Anderwerty) [inline src/test/java/MyStackTest.java]: `"String"` looks like a magic number. Please use a constant.

File: src/test/java/MyLinkedListTestAbstract.java [hunk 31]
```diff
@@ -0,0 +1,17 @@
...
+
+    public abstract T createSampleValue();
+    public abstract MyLinkedList<T> createSampleMyLinkedList();
+    public abstract List<T> createSampleList();
+    public abstract T [] dataForList(int sizeArray);
+    public abstract T createRandomValue();
+    public abstract void fillArray (T [] array);
+
+
+}
```
Comment (Anderwerty) [inline src/test/java/MyLinkedListTestAbstract.java]: Please add a blank line at the end of the file.

File: src/test/java/MyLinkedListTest.java [hunk 32]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    @Override
+    public void fillArray(Integer[] array) {
+        for (int i = 0; i < array.length; i++) {
+            array[i] = i;
+        }
+    }
+
+    private int randomInt(int limit) {
```
Comment (Anderwerty) [inline src/test/java/MyLinkedListTest.java:136]: How do you use this method?

File: src/test/java/MyArrayListTest.java [hunk 33]
```diff
@@ -0,0 +1,98 @@
...
+
+    @Override
+    public Integer[] dataForList(int sizeArray) {
+        Integer[] result = new Integer[sizeArray];
+        fillArray(result);
+        return result;
+    }
+
+    @Override
+    public Integer createRandomValue() {
```
Comment (Anderwerty) [inline src/test/java/MyArrayListTest.java:86]: Test data should be static, not dynamic.

**Review by Anderwerty — CHANGES_REQUESTED** [review]
You make test code so difficult to read and support. It better to create separate test classes.

File: src/main/java/MyArrayList.java (hunk 6 above)
Comment (Bronzza) [reply src/main/java/MyArrayList.java]: What does it mean?  Is it about the method "checkIsExpand" or about toString?

Comment (Anderwerty) [reply src/main/java/MyArrayList.java]: Look where you use checklsExpand. First, you should check if there any empty place in the array to put new element. If no make your array bigger.

**Review by Anderwerty — CHANGES_REQUESTED** [review]

File: src/main/java/MyStackInterface.java (hunk 19 above)
Comment (Bronzza) [reply src/main/java/MyStackInterface.java]: What means the same here, "Remove redundant blank line"?

File: .idea/uiDesigner.xml (hunk 1 above)
Comment (Bronzza) [reply .idea/uiDesigner.xml]: Couse of my gitIgnore I cant add it to the repository, how should I add this file?

Comment (Anderwerty) [reply .idea/uiDesigner.xml]: Maybe first commit this file and then commit gitignore. Remove this file from the repo.

File: src/main/java/MyArrayList.java (hunk 7 above)
Comment (Bronzza) [reply src/main/java/MyArrayList.java]: Should I use reflection in test method to assign value to counter more than to DEFAULT_CAPACITY? Couse in other case this situation will never happen.

File: src/main/java/MyArrayList.java (hunk 10 above)
Comment (Bronzza) [reply src/main/java/MyArrayList.java]: In this case in output will see null objects wich are stored in inner array of the class

File: src/main/java/MyLinkedList.java (hunk 12 above)
Comment (Bronzza) [reply src/main/java/MyLinkedList.java]: There are two tests in "@Test add1()" for invalid index, for negative index and for index that is out of list size range, which both expect indexOutOfBoundExceptions

File: src/main/java/MyArrayList.java (hunk 7 above)
Comment (Anderwerty) [reply src/main/java/MyArrayList.java]: no, you should not.

File: src/test/java/MyStackTest.java (hunk 26 above)
Comment (Bronzza) [reply src/test/java/MyStackTest.java]: Pop moves you to remove in MyLinkedList which moves you to method “unlink” from there to method “FindElemnt” which throws an exception which is cathced by assert -> works this way. Removing "(expected = IndexOutOfBoundsException.class)"text from the annotation will lead to fail the test

File: src/main/java/MyLinkedList.java (hunk 12 above)
Comment (Anderwerty) [reply src/main/java/MyLinkedList.java]: ok

File: src/main/java/MyStackInterface.java (hunk 19 above)
Comment (Anderwerty) [reply src/main/java/MyStackInterface.java]: No put all classes into some root package

File: src/main/java/MyArrayList.java (hunk 6 above)
Comment (Bronzza) [reply src/main/java/MyArrayList.java]: public void add(T t) {
        array[counter] = t;
        counter++;
In this method counter is incremented after addition, so it point out on next field of array (after last added element). That is why we check counter and size for "==". Method works.

**Review by Anderwerty — CHANGES_REQUESTED** [review]

File: src/test/java/MyArrayListTest.java [hunk 34]
```diff
@@ -0,0 +1,97 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Stream;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+
+public class MyArrayListTest extends MyArrayListTestAbs<Integer> {
+    private int indexTest;
```
Comment (Anderwerty) [inline src/test/java/MyArrayListTest.java]: Why not constant here?

File: src/test/java/MyStackTest.java (hunk 26 above)
Comment (Anderwerty) [reply src/test/java/MyStackTest.java]: assert does not catch an exception

**Review by Anderwerty — CHANGES_REQUESTED** [review]

**Review by Anderwerty — APPROVED** [review]
Remove files that should not be in a repo.

**Pull request merged**